File: analysis/factor_analysis/efa_analysis.py
import pandas as pd
import numpy as np
from factor_analyzer import FactorAnalyzer
from factor_analyzer.factor_analyzer import calculate_kmo, calculate_bartlett_sphericity
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import os

# Create results directory
RESULTS_DIR = 'efa_results'
os.makedirs(RESULTS_DIR, exist_ok=True)

# Read the data
print("Reading Excel file...")
df = pd.read_excel('../../data/raw/dat_EFA_26.xlsx')

# Print data info
print("\nDataset Info:")
print(df.info())

print("\nColumn names:")
print(df.columns.tolist())

def prepare_data(df):
    """Prepare data for EFA"""
    # First, identify numeric columns that could be items
    numeric_cols = df.select_dtypes(include=[np.number]).columns
    print(f"\nNumeric columns: {len(numeric_cols)}")
    print(numeric_cols.tolist())
    
    # Remove any missing values
    data_clean = df[numeric_cols].dropna()
    print(f"\nShape after cleaning: {data_clean.shape}")
    
    # Check for low variance columns
    variances = data_clean.var()
    low_var_cols = variances[variances < 0.001].index
    if len(low_var_cols) > 0:
        print(f"\nRemoving {len(low_var_cols)} low variance columns:")
        print(low_var_cols.tolist())
        data_clean = data_clean.drop(columns=low_var_cols)
    
    # Standardize the data
    data_scaled = stats.zscore(data_clean)
    return pd.DataFrame(data_scaled, columns=data_clean.columns)

def check_efa_suitability(data):
    """Check if data is suitable for EFA using KMO and Bartlett's test"""
    try:
        print("\nChecking EFA suitability...")
        print(f"Data shape: {data.shape}")
        
        # Calculate KMO
        kmo_all, kmo_model = calculate_kmo(data)
        print("\nKMO scores:")
        print(f"Overall KMO: {kmo_model:.3f}")
        print("\nKMO scores for each variable:")
        for var, kmo in zip(data.columns, kmo_all):
            print(f"{var}: {kmo:.3f}")
        
        # Calculate Bartlett's test
        chi_square_value, p_value = calculate_bartlett_sphericity(data)
        print("\nBartlett's test of sphericity:")
        print(f"Chi-square: {chi_square_value:.3f}")
        print(f"p-value: {p_value:.3e}")
        
        return kmo_model, p_value
    except Exception as e:
        print(f"Error in checking EFA suitability: {str(e)}")
        return None, None

def determine_n_factors(data):
    """Determine optimal number of factors using multiple methods"""
    try:
        print("\nDetermining number of factors...")
        print(f"Data shape: {data.shape}")
        
        # Initialize factor analyzer
        fa = FactorAnalyzer(rotation=None, n_factors=data.shape[1])
        fa.fit(data)
        
        # Scree plot
        ev, v = fa.get_eigenvalues()
        plt.figure(figsize=(10, 6))
        plt.plot(range(1, len(ev) + 1), ev)
        plt.title('Scree Plot')
        plt.xlabel('Factors')
        plt.ylabel('Eigenvalue')
        plt.axhline(y=1, color='r', linestyle='--')
        plt.savefig(os.path.join(RESULTS_DIR, 'scree_plot.png'))
        plt.close()
        
        # Parallel Analysis
        n_iterations = 1000
        n_vars = data.shape[1]
        n_samples = data.shape[0]
        
        # Generate random eigenvalues
        random_evals = np.zeros((n_iterations, n_vars))
        for i in range(n_iterations):
            random_data = np.random.normal(size=(n_samples, n_vars))
            fa_random = FactorAnalyzer(rotation=None, n_factors=n_vars)
            fa_random.fit(random_data)
            random_evals[i, :], _ = fa_random.get_eigenvalues()
        
        # Get 95th percentile of random eigenvalues
        percentile_95 = np.percentile(random_evals, 95, axis=0)
        
        # Plot parallel analysis
        plt.figure(figsize=(10, 6))
        plt.plot(range(1, len(ev) + 1), ev, 'b', label='Actual Data')
        plt.plot(range(1, len(percentile_95) + 1), percentile_95, 'r--', label='95th Percentile of Random Data')
        plt.title('Parallel Analysis')
        plt.xlabel('Factors')
        plt.ylabel('Eigenvalue')
        plt.legend()
        plt.savefig(os.path.join(RESULTS_DIR, 'parallel_analysis.png'))
        plt.close()
        
        # Determine number of factors
        n_factors = sum(ev > percentile_95)
        print(f"\nSuggested number of factors from parallel analysis: {n_factors}")
        
        # Save eigenvalues to file
        with open(os.path.join(RESULTS_DIR, 'eigenvalues.txt'), 'w') as f:
            f.write("Eigenvalues:\n")
            for i, eigenvalue in enumerate(ev, 1):
                f.write(f"Factor {i}: {eigenvalue:.3f}\n")
            f.write("\nCumulative variance explained:\n")
            cumulative = np.cumsum(ev) / sum(ev) * 100
            for i, cum_var in enumerate(cumulative, 1):
                f.write(f"Factor 1-{i}: {cum_var:.1f}%\n")
        
        return n_factors
    except Exception as e:
        print(f"Error in determining number of factors: {str(e)}")
        return None

def perform_efa(data, n_factors):
    """Perform EFA with different rotation methods"""
    try:
        print(f"\nPerforming EFA with {n_factors} factors...")
        print(f"Data shape: {data.shape}")
        
        rotations = ['varimax', 'promax', 'oblimin']
        results = {}
        
        for rotation in rotations:
            print(f"\nPerforming {rotation} rotation...")
            # Initialize and fit factor analyzer
            fa = FactorAnalyzer(rotation=rotation, n_factors=n_factors)
            fa.fit(data)
            
            # Get loadings
            loadings = pd.DataFrame(
                fa.loadings_,
                columns=[f'Factor{i+1}' for i in range(n_factors)],
                index=data.columns
            )
            
            # Get variance explained
            var_explained = fa.get_factor_variance()
            variance = pd.DataFrame(
                var_explained,
                columns=['SS Loadings', 'Proportion Var', 'Cumulative Var'],
                index=[f'Factor{i+1}' for i in range(n_factors)]
            )
            
            if rotation in ['promax', 'oblimin']:
                # Get factor correlations for oblique rotations
                factor_corr = pd.DataFrame(
                    fa.corr_,
                    columns=[f'Factor{i+1}' for i in range(n_factors)],
                    index=[f'Factor{i+1}' for i in range(n_factors)]
                )
            else:
                factor_corr = None
            
            results[rotation] = {
                'loadings': loadings,
                'variance': variance,
                'factor_corr': factor_corr
            }
            
            # Create heatmap of factor loadings
            plt.figure(figsize=(12, 8))
            sns.heatmap(loadings, annot=True, cmap='RdBu', center=0, fmt='.3f')
            plt.title(f'Factor Loadings ({rotation} rotation)')
            plt.tight_layout()
            plt.savefig(os.path.join(RESULTS_DIR, f'loadings_{rotation}.png'))
            plt.close()
            
            # Save results to file
            with open(os.path.join(RESULTS_DIR, f'efa_results_{rotation}.txt'), 'w') as f:
                f.write(f"EFA Results - {rotation} rotation\n")
                f.write("=" * 50 + "\n\n")
                f.write("Factor Loadings:\n")
                f.write(loadings.to_string())
                f.write("\n\nVariance Explained:\n")
                f.write(variance.to_string())
                if factor_corr is not None:
                    f.write("\n\nFactor Correlations:\n")
                    f.write(factor_corr.to_string())
                
                # Add factor interpretation
                f.write("\n\nFactor Interpretation:\n")
                f.write("-" * 30 + "\n")
                for factor in range(n_factors):
                    f.write(f"\nFactor {factor + 1}:\n")
                    # Get items with significant loadings (> 0.4)
                    sig_items = loadings[loadings[f'Factor{factor+1}'].abs() > 0.4]
                    sig_items = sig_items.sort_values(by=f'Factor{factor+1}', key=abs, ascending=False)
                    f.write(f"Significant items (|loading| > 0.4):\n")
                    for item, loading in sig_items[f'Factor{factor+1}'].items():
                        f.write(f"{item}: {loading:.3f}\n")
        
        return results
    except Exception as e:
        print(f"Error in performing EFA: {str(e)}")
        return None

def main():
    try:
        # Prepare data
        print("Preparing data...")
        data = prepare_data(df)
        print(f"Analyzing {len(data.columns)} variables with {len(data)} complete cases")
        
        # Check suitability
        kmo, bartlett_p = check_efa_suitability(data)
        
        if kmo is not None and bartlett_p is not None:
            if kmo < 0.5 or bartlett_p > 0.05:
                print("\nWarning: Data may not be suitable for EFA!")
                print(f"KMO: {kmo:.3f} (should be > 0.5)")
                print(f"Bartlett's p-value: {bartlett_p:.3e} (should be < 0.05)")
                if input("Continue anyway? (y/n): ").lower() != 'y':
                    return
        
        # Determine number of factors
        n_factors = determine_n_factors(data)
        
        if n_factors is not None and n_factors > 0:
            # Perform EFA
            results = perform_efa(data, n_factors)
            
            if results is not None:
                print("\nAnalysis complete! Results have been saved to the 'efa_results' directory.")
            else:
                print("\nError occurred during EFA. Please check the error messages above.")
        else:
            print("\nError occurred while determining number of factors. Please check the error messages above.")
    
    except Exception as e:
        print(f"\nAn error occurred: {str(e)}")

if __name__ == "__main__":
    main() 

File: analysis/factor_analysis/efa_analysis.R
# Load required packages
library(readxl)      # For reading Excel files
library(psych)       # For factor analysis
library(GPArotation) # For factor rotation
library(corrplot)    # For correlation visualization
library(nFactors)    # For determining number of factors
library(tidyverse)   # For data manipulation
library(ggplot2)     # For enhanced plotting
library(papaja)      # For APA style

# Create results directory
dir.create("efa_results", showWarnings = FALSE)

# Read the data
cat("Reading data...\n")
data <- read_excel("../../data/raw/dat_EFA_26.xlsx")

# Print initial data information
cat("\nInitial data structure:\n")
str(data)
cat("\nColumn names:\n")
print(colnames(data))

# Function to prepare data for EFA
prepare_data <- function(data) {
  # Select numeric columns
  numeric_data <- data %>% select_if(is.numeric)
  
  cat("\nNumeric columns selected:", ncol(numeric_data), "\n")
  cat("Numeric column names:\n")
  print(colnames(numeric_data))
  
  # Check for missing values
  missing_summary <- colSums(is.na(numeric_data))
  cat("\nMissing values per column:\n")
  print(missing_summary)
  
  # Remove columns with near-zero variance
  var_threshold <- 0.001
  vars <- apply(numeric_data, 2, var, na.rm = TRUE)
  cat("\nVariance per column:\n")
  print(vars)
  
  numeric_data <- numeric_data[, vars > var_threshold]
  
  # Handle missing values
  complete_data <- na.omit(numeric_data)
  
  cat("\nDimensions after cleaning:\n")
  cat("Original:", nrow(data), "rows,", ncol(data), "columns\n")
  cat("After cleaning:", nrow(complete_data), "rows,", ncol(complete_data), "columns\n")
  
  # Standardize the data
  scaled_data <- scale(complete_data)
  
  # Basic checks on scaled data
  cat("\nScaled data summary:\n")
  print(summary(scaled_data))
  
  return(list(
    data = scaled_data,
    removed_cols = names(vars[vars <= var_threshold]),
    complete_cases = nrow(complete_data)
  ))
}

# Function to check EFA suitability
check_efa_suitability <- function(data) {
  # Correlation matrix
  cor_matrix <- cor(data)
  
  # Save correlation matrix
  pdf("efa_results/correlation_matrix.pdf", width = 12, height = 12)
  corrplot(cor_matrix, method = "color", type = "upper", 
           tl.col = "black", tl.srt = 45, tl.cex = 0.7)
  dev.off()
  
  # KMO test
  kmo <- KMO(cor_matrix)
  
  # Bartlett's test
  bartlett <- cortest.bartlett(cor_matrix, n = nrow(data))
  
  # Save results
  sink("efa_results/suitability_tests.txt")
  cat("EFA Suitability Tests\n")
  cat("=====================\n\n")
  cat("KMO Test Results:\n")
  print(kmo)
  cat("\nBartlett's Test Results:\n")
  print(bartlett)
  sink()
  
  return(list(kmo = kmo$MSA, bartlett = bartlett))
}

# Function to determine number of factors
determine_factors <- function(data) {
  # Eigenvalues
  ev <- eigen(cor(data))
  
  cat("\nEigenvalues:\n")
  print(ev$values)
  
  # Parallel analysis
  pa <- parallel(subject = nrow(data), var = ncol(data),
                rep = 1000, cent = 0.95)
  
  # Scree plot with parallel analysis
  pdf("efa_results/scree_plot.pdf")
  scree(cor(data), factors = FALSE)
  abline(h = 1, col = "red", lty = 2)  # Kaiser criterion
  points(pa$eigen$qevpea, type = "l", col = "blue", lty = 2)  # Parallel analysis
  legend("topright", 
         legend = c("Actual", "Parallel Analysis", "Kaiser Criterion"),
         col = c("black", "blue", "red"),
         lty = c(1, 2, 2))
  dev.off()
  
  # VSS analysis
  vss_result <- VSS(data, rotate = "varimax")
  
  # Save results
  sink("efa_results/factor_determination.txt")
  cat("Factor Number Determination\n")
  cat("=========================\n\n")
  cat("Eigenvalues:\n")
  print(ev$values)
  cat("\nKaiser criterion suggestion:", sum(ev$values > 1), "\n")
  cat("Parallel Analysis Suggestion:", sum(ev$values > pa$eigen$qevpea), "\n")
  cat("\nVSS Analysis Results:\n")
  print(vss_result)
  sink()
  
  # Return 7 factors as requested
  return(7)
}

# Function to create APA-style scree plot
create_scree_plot <- function(data, pa_result) {
  # Calculate eigenvalues
  ev <- eigen(cor(data))
  
  # Create data frame for plotting
  n_factors <- length(ev$values)
  plot_data <- data.frame(
    Factor = 1:n_factors,
    Empirical = ev$values,
    Parallel = c(pa_result$eigen$qevpea, rep(NA, n_factors - length(pa_result$eigen$qevpea)))
  )
  
  # Create ggplot with APA style
  p <- ggplot(plot_data, aes(x = Factor)) +
    geom_line(aes(y = Empirical, linetype = "Empirical Eigenvalues"), 
              linewidth = 0.5) +
    geom_point(aes(y = Empirical), size = 2) +
    geom_line(aes(y = Parallel, linetype = "Parallel Analysis"), 
              linewidth = 0.5) +
    geom_hline(yintercept = 1, linetype = "dotted", linewidth = 0.5) +
    scale_linetype_manual(name = NULL,
                         values = c("Empirical Eigenvalues" = "solid", 
                                  "Parallel Analysis" = "dashed")) +
    theme_apa() +
    theme(
      legend.position = "bottom",
      legend.box.spacing = unit(0.5, "lines"),
      axis.text = element_text(size = 10),
      axis.title = element_text(size = 12)
    ) +
    labs(
      x = "Factor Number",
      y = "Eigenvalue",
      caption = "Note. Dotted line represents Kaiser criterion (eigenvalue = 1)."
    )
  
  # Save plot
  ggsave("efa_results/scree_plot_apa.pdf", p, 
         width = 6.5, height = 5, device = cairo_pdf)
}

# Function to create APA-style correlation heatmap
create_correlation_heatmap <- function(cor_matrix) {
  pdf("efa_results/correlation_matrix_apa.pdf", 
      width = 8.5, height = 8)
  corrplot(cor_matrix,
           method = "color",
           type = "upper",
           tl.col = "black",
           tl.srt = 45,
           tl.cex = 0.7,
           cl.cex = 0.7,
           addCoef.col = "black",
           number.cex = 0.5,
           col = colorRampPalette(c("#000000", "#FFFFFF"))(200),
           diag = FALSE,
           title = "Inter-Item Correlation Matrix",
           mar = c(0,0,1.5,0))
  dev.off()
}

# Function to create APA-style factor loading plot
create_factor_loading_plot <- function(fa_result, rotation_name) {
  # Create loading matrix with only significant loadings (> 0.3)
  loadings <- fa_result$loadings
  loadings[abs(loadings) < 0.3] <- NA
  
  # Convert to data frame
  load_df <- as.data.frame(unclass(loadings))
  load_df$Item <- rownames(loadings)
  
  # Reshape for ggplot
  load_long <- load_df %>%
    gather(Factor, Loading, -Item) %>%
    filter(!is.na(Loading))
  
  # Create plot with APA style
  p <- ggplot(load_long, aes(x = Factor, y = Item)) +
    geom_tile(aes(fill = Loading), color = "white") +
    scale_fill_gradient2(low = "white", mid = "gray80", high = "black",
                        midpoint = 0, limits = c(-1, 1),
                        name = "Factor\nLoading") +
    theme_apa() +
    theme(
      axis.text.x = element_text(angle = 0, size = 10),
      axis.text.y = element_text(size = 8),
      legend.position = "right",
      legend.key.size = unit(0.8, "lines"),
      plot.caption = element_text(size = 8, hjust = 0)
    ) +
    labs(
      title = paste("Factor Loadings -", rotation_name, "Rotation"),
      x = "Factor",
      y = "Item",
      caption = "Note. Only loadings > |0.30| are shown. Darker shading indicates stronger positive loadings."
    )
  
  # Save plot
  ggsave(paste0("efa_results/loadings_", tolower(rotation_name), "_apa.pdf"),
         p, width = 8.5, height = 11, device = cairo_pdf)
  
  # Create additional table format for publication
  loadings_table <- round(unclass(loadings), 3)
  loadings_table[abs(loadings_table) < 0.3] <- NA
  write.csv(loadings_table, 
            paste0("efa_results/loadings_", tolower(rotation_name), "_table.csv"))
}

# Function to create factor correlation table (for oblique rotations)
create_factor_correlation_table <- function(fa_result, rotation_name) {
  if (!rotation_name == "Varimax") {
    factor_cors <- round(fa_result$Phi, 3)
    write.csv(factor_cors,
              paste0("efa_results/factor_correlations_", tolower(rotation_name), ".csv"))
  }
}

# Function to compare different factor solutions
compare_factor_solutions <- function(data, max_factors = 8) {
  results <- data.frame(
    Factors = 3:max_factors,
    RMSEA = NA,
    TLI = NA,
    BIC = NA,
    ChiSq = NA,
    df = NA,
    p_value = NA
  )
  
  for(i in 3:max_factors) {
    fa_result <- fa(data, nfactors = i, rotate = "varimax", fm = "ml")
    results$RMSEA[i-2] <- fa_result$RMSEA[1]
    results$TLI[i-2] <- fa_result$TLI
    results$BIC[i-2] <- fa_result$BIC
    results$ChiSq[i-2] <- fa_result$chi
    results$df[i-2] <- fa_result$dof
    results$p_value[i-2] <- fa_result$PVAL
  }
  
  # Save comparison results
  write.csv(results, "efa_results/factor_comparison.csv")
  
  # Create comparison plot
  p <- results %>%
    gather(Metric, Value, -Factors) %>%
    filter(Metric %in% c("RMSEA", "TLI")) %>%
    ggplot(aes(x = Factors, y = Value, group = Metric)) +
    geom_line() +
    geom_point() +
    facet_wrap(~Metric, scales = "free_y") +
    theme_apa() +
    labs(
      x = "Number of Factors",
      y = "Fit Index Value",
      caption = "Note. RMSEA = Root Mean Square Error of Approximation; TLI = Tucker-Lewis Index."
    )
  
  ggsave("efa_results/factor_comparison_plot.pdf", p, 
         width = 8, height = 4, device = cairo_pdf)
  
  return(results)
}

# Main analysis function
main <- function() {
  # Read and prepare data
  cat("Reading data...\n")
  data <- read_excel("../../data/raw/dat_EFA_26.xlsx")
  
  # Prepare data
  prep_result <- prepare_data(data)
  prepared_data <- prep_result$data
  
  # Compare different factor solutions
  cat("\nComparing different factor solutions...\n")
  comparison_results <- compare_factor_solutions(prepared_data)
  print(comparison_results)
  
  # Create publication-quality plots
  pa_result <- parallel(subject = nrow(prepared_data), 
                       var = ncol(prepared_data),
                       rep = 1000, 
                       cent = 0.95)
  
  create_scree_plot(prepared_data, pa_result)
  create_correlation_heatmap(cor(prepared_data))
  
  # Perform EFA with different rotations
  rotations <- c("Varimax", "Promax", "Oblimin")
  for(rotation in rotations) {
    fa_result <- fa(prepared_data, 
                    nfactors = 7, 
                    rotate = tolower(rotation), 
                    fm = "ml")
    create_factor_loading_plot(fa_result, rotation)
    create_factor_correlation_table(fa_result, rotation)
  }
  
  # Create summary statistics table
  summary_stats <- data.frame(
    Factor = 1:7,
    Eigenvalue = eigen(cor(prepared_data))$values[1:7],
    Variance_Explained = fa_result$Vaccounted[2,],
    Cumulative_Variance = fa_result$Vaccounted[3,]
  )
  write.csv(round(summary_stats, 3),
            "efa_results/factor_summary_statistics.csv")
  
  cat("\nAnalysis complete! Results have been saved to the 'efa_results' directory.\n")
}

# Run the analysis
main() 